crates/polars-ops/src/chunked_array/datetime/replace_time_zone.rs

```diff
@@ -82,3 +85,101 @@ pub fn replace_time_zone(
+        TimeUnit::Milliseconds => timestamp_ms_to_datetime,
+        TimeUnit::Microseconds => timestamp_us_to_datetime,
+        TimeUnit::Nanoseconds => timestamp_ns_to_datetime,
+    };
+    let datetime_to_timestamp: fn(NaiveDateTime) -> i64 = match datetime.time_unit() {
+        TimeUnit::Milliseconds => datetime_to_timestamp_ms,
+        TimeUnit::Microseconds => datetime_to_timestamp_us,
+        TimeUnit::Nanoseconds => datetime_to_timestamp_ns,
+    };
+    let out: Result<ChunkedArray<Int64Type>, PolarsError> = match convert_tz.len() {
+        1 => match unsafe { convert_tz.get_unchecked(0) } {
+            Some(convert_tz) => {
+                let to_tz = parse_time_zone(convert_tz)?;
+                Ok(datetime.0.apply(|timestamp_opt| {
+                    timestamp_opt.map(|ts| {
+                        let ndt = timestamp_to_datetime(ts);
+                        datetime_to_timestamp(naive_utc_to_naive_local_in_new_time_zone(
+                            &from_tz, &to_tz, ndt,
+                        ))
+                    })
+                }))
+            },
+            _ => Ok(datetime.0.apply(|_| None)),
+        },
+        _ => try_binary_elementwise(
+            datetime,
+            convert_tz,
+            |timestamp_opt, convert_tz_opt| match (timestamp_opt, convert_tz_opt) {
+                (Some(timestamp), Some(convert_tz)) => {
+                    let ndt = timestamp_to_datetime(timestamp);
+                    let to_tz = parse_time_zone(convert_tz)?;
+                    Ok(Some(datetime_to_timestamp(
+                        naive_utc_to_naive_local_in_new_time_zone(&from_tz, &to_tz, ndt),
+                    )))
+                },
+                _ => Ok(None),
+            },
+        ),
+    };
+    let out = out?.into_datetime(datetime.time_unit(), None);
+    Ok(out)
+}
+
+pub fn from_local_datetime(
+    datetime: &Logical<DatetimeType, Int64Type>,
+    naive_tz: &Utf8Chunked,
+    out_tz: &str,
+    ambiguous: &str,
+) -> PolarsResult<DatetimeChunked> {
+    let to_tz = parse_time_zone(out_tz)?;
+    let ambig = Ambiguous::from_str(ambiguous)?;
+    let timestamp_to_datetime: fn(i64) -> NaiveDateTime = match datetime.time_unit() {
+        TimeUnit::Milliseconds => timestamp_ms_to_datetime,
+        TimeUnit::Microseconds => timestamp_us_to_datetime,
+        TimeUnit::Nanoseconds => timestamp_ns_to_datetime,
+    };
+    let datetime_to_timestamp: fn(NaiveDateTime) -> i64 = match datetime.time_unit() {
+        TimeUnit::Milliseconds => datetime_to_timestamp_ms,
+        TimeUnit::Microseconds => datetime_to_timestamp_us,
+        TimeUnit::Nanoseconds => datetime_to_timestamp_ns,
+    };
+    let out = match naive_tz.len() {
+        1 => match unsafe { naive_tz.get_unchecked(0) } {
+            Some(naive_tz) => datetime.0.try_apply(|timestamp| {
+                let ndt = timestamp_to_datetime(timestamp);
+                let from_tz = parse_time_zone(naive_tz)?;
```

Comment:
does this need to be parsed for each element? If it's just a single one, I'd suggest putting this one outside the `try_apply` as well